Kernel/Storage/NVME/NVMEController.cpp

```diff
@@ -0,0 +1,315 @@
+{
+
+    RefPtr<Memory::PhysicalPage> prp_dma_buffer;
+    OwnPtr<Memory::Region> prp_dma_region;
+    auto namespace_data_struct = ByteBuffer::create_zeroed(NVME_IDENTIFY_SIZE).release_value();
+    u32 active_namespace_list[NVME_IDENTIFY_SIZE / sizeof(u32)];
+
+    if (auto buffer = MM.dma_allocate_buffer(NVME_IDENTIFY_SIZE, "Identify PRP", Memory::Region::Access::ReadWrite, prp_dma_buffer); buffer.is_error()) {
+        dmesgln("Failed to allocate memory for ADMIN CQ queue");
+        // TODO:Need to figure out better error propagation
+        VERIFY_NOT_REACHED();
+    } else {
+        prp_dma_region = buffer.release_value();
+    }
+    // Get the active namespace
+    {
+        nvme_submission sub {};
+        u16 status = 0;
+        sub.op = OP_ADMIN_IDENTIFY;
+        sub.data_ptr.prp1 = reinterpret_cast<u64>(AK::convert_between_host_and_little_endian(prp_dma_buffer->paddr().as_ptr()));
+        sub.cdw10 = NVME_CNS_ID_ACTIVE_NS & 0xff;
+        status = submit_admin_command(sub, true);
+        if (status) {
+            dmesgln("Failed to identify active namespace command");
+            VERIFY_NOT_REACHED();
+        }
+        if (void* fault_at; !safe_memcpy(active_namespace_list, prp_dma_region->vaddr().as_ptr(), NVME_IDENTIFY_SIZE, fault_at)) {
+            VERIFY_NOT_REACHED();
+        }
+    }
+    // Get the NAMESPACE attributes
+    {
+        nvme_submission sub {};
+        u16 status = 0;
+        for (auto nsid : active_namespace_list) {
+            memset(prp_dma_region->vaddr().as_ptr(), 0, NVME_IDENTIFY_SIZE);
+            // Invalid NS
+            if (nsid == 0)
+                break;
+            sub.op = OP_ADMIN_IDENTIFY;
+            sub.data_ptr.prp1 = reinterpret_cast<u64>(AK::convert_between_host_and_little_endian(prp_dma_buffer->paddr().as_ptr()));
+            sub.cdw10 = NVME_CNS_ID_NS & 0xff;
+            sub.nsid = nsid;
+            status = submit_admin_command(sub, true);
+            if (status) {
+                dmesgln("Failed identify namespace with nsid {}", nsid);
+                VERIFY_NOT_REACHED();
+            }
+            if (void* fault_at; !safe_memcpy(namespace_data_struct.data(), prp_dma_region->vaddr().as_ptr(), NVME_IDENTIFY_SIZE, fault_at)) {
+                VERIFY_NOT_REACHED();
+            }
+            auto val = get_ns_features(namespace_data_struct);
+            auto block_counts = val.get<0>();
+            auto block_size = 1 << val.get<1>();
+
+            dbgln_if(NVME_DEBUG, "NVMe: Block count is {} and Block size is {}", block_counts, block_size);
+
+            m_namespaces.append(NVMENameSpace::create(m_queues, nsid, val.get<0>(), block_size));
+            m_device_count++;
+            dbgln_if(NVME_DEBUG, "NVMe: Initialized namespace with NSID: {}", nsid);
+        }
+    }
+}
+Tuple<u64, u8> NVMEController::get_ns_features(ByteBuffer& identify_data_struct)
+{
+    auto flbas = identify_data_struct[26] & 0xf;
```

Comment:
I see a lot of magic numbers around here, either adding comments to the spec or pulling them out into constants and commenting those at the top of the file I would consider useful

Reply:
Thanks for the comment. I tried to avoid them but some slipped through. I will add these as constants in the next revision.